Editorial review: Document new web speech api features (#41145)
mdn/content: files/en-us/web/api/web_speech_api/using_the_web_speech_api/index.md

---

{{DefaultAPISidebar("Web Speech API")}}

The Web Speech API provides two distinct areas of functionality — speech recognition, and speech synthesis (also known as text to speech, or TTS) — which open up interesting new possibilities for accessibility, and control mechanisms. This article provides an introduction to both areas, along with demos.

## Speech recognition

Speech recognition involves receiving speech through a device's microphone (or an audio track), which is then checked by a speech recognition service. When a word or phrase is successfully recognized, it is returned as a result (or list of results) as a text string, and further actions can be initiated as a result.

The Web Speech API has a main controller interface for this — {{domxref("SpeechRecognition")}} — plus several related interfaces for representing results, etc.

> Generally, the default speech recognition system available on the device will be used for the speech recognition — most modern OSes have a speech recognition system for issuing voice commands. Think about Dictation on macOS or Cortana on Windows. On some browsers, such as Chrome, using Speech Recognition on a web page involves a server-based recognition engine. Your audio is sent to a web service for recognition processing, so it won't work offline.

Reviewer: Chrome recently launched on-device Web Speech, so it would be more accurate to say that speech recognition "may" involve a server-based recognition. Or it might be better to just use a different example like Safari that always uses server-side speech recognition.

> Generally, the default speech recognition system available on the device will be used for the speech recognition — most modern OSes have a speech recognition system for issuing voice commands. Think about Dictation on macOS or Cortana on Windows. On some browsers, such as Chrome, using Speech Recognition on a web page involves a server-based recognition engine. Your audio is sent to a web service for recognition processing, so it won't work offline.

Author: I've replaced Cortana with Copilot — shows how long ago these docs were originally written!

I've restructured this bit to avoid the server-based versus on-device issue you highlighted. Let me know what you think.

The Web Speech API has a main controller interface for this — {{domxref("SpeechRecognition")}} — plus several related interfaces for representing results, etc.

Generally, the default speech recognition system available on the device will be used for the speech recognition. Most modern OSes have a speech recognition system for issuing voice commands such as Dictation on macOS or Copilot on Windows.

By default, using Speech Recognition on a web page usually involves a server-based recognition engine. Your audio is sent to a web service for recognition processing, so it won't work offline.

To improve privacy and performance, it is possible to specify that you want the speech recognition performed on-device, thereby ensuring that neither audio nor transcribed speech are sent to a third-party service for processing. We specifically cover the on-device functionality in [on-device speech recognition](#on-device_speech_recognition).

### Demo

> To demonstrate usage of Web speech recognition, we've written a demo called [Speech color changer](https://github.com/mdn/dom-examples/tree/main/web-speech-api/speech-color-changer). When the button is pressed, you can say an HTML color keyword, and the app's background color will change to that color.

Author: updated

  "coral",
  // …
];
```

#### Creating a speech recognition instance

> The next thing to do is define a speech recognition instance to control the recognition for our application. This is done using the {{domxref("SpeechRecognition.SpeechRecognition()","SpeechRecognition()")}} constructor.

Reviewer 2: ```suggestion
Next, we define a speech recognition instance to control the recognition in our application. This is done using the {{domxref("SpeechRecognition.SpeechRecognition()","SpeechRecognition()")}} constructor.
```

Author: updated

#### Starting the speech recognition

> After grabbing references to the output paragraph, the `<html>` element (so we can output diagnostic messages and update the app background color later on), the instructions paragraph, and the `<button>`, we implement an `onclick` handler so that when it is pressed, the speech recognition service will start. This is achieved by calling {{domxref("SpeechRecognition.start()")}}. A `forEach()` method is also used, to output colored indicators showing what colors to try saying.

Author: applied

- If the resulting value is `available`, it means that the language pack is available locally, so recognition can begin. In this case, we run `start()` and log a message to the console when the app is ready to receive speech.
- If the value is something else (`downloadable` or `downloading`), we print a diagnostic message to inform the user that a language code download is commencing, then run the `install()` method to handle the download.

The `install()` method works in a similar way to the `available()` method, except that its options object only takes the `langs` array. When run, it starts downloading all the language packs for the languages indicated in `langs` and returns a {{jsxref("Promise")}} that resolves with a boolean indicating whether the specified language packs were downloaded and installed successfully (`true`) or not (`false`).

In this basic demo, we just write a diagnostic message to indicate the success and failure cases; in a more substantial demo you'd probably want to disable the controls while the download process occurs, and enable them again when the promise resolves.

### Permissions-policy integration

> Usage of the `available()` and `install()` methods is controlled by the {{httpheader("Permissions-Policy/on-device-speech-recognition", "on-device-speech-recognition")}} {{httpheader("Permissions-Policy")}} directive. Specifically, where a defined policy blocks usage, any attempts to call these methods will fail.
> 
> The default allowlist value for `on-device-speech-recognition` is `self`, so you don't need to worry about this unless you are attempting to use these features in embedded cross-origin documents, or specifically want to disable their usage.

Reviewer 2: removing some extra words to shorten and some pronoun replacements for clarity
```suggestion
The use of the `available()` and `install()` methods is controlled by the {{httpheader("Permissions-Policy/on-device-speech-recognition", "on-device-speech-recognition")}} directive of the {{httpheader("Permissions-Policy")}} response header. If a policy blocks their use, any attempts to call these methods will fail.

The default allowlist value for `on-device-speech-recognition` is `self`. This means you don't need to worry about adjusting the policy unless you're attempting to use these methods in embedded cross-origin documents or want to explicitly disable their use.
```

### Unprefixed Web Speech API

> In the original speech color changer demo, we include lines to handle the issue that some browsers still support the Web Speech API with prefixed properties (see the [Prefixed properties](#prefixed_properties) section for more details).
> 
> In the on-device version, we don't include the prefix-handling code because the new implementations that support this functionality do so without prefixes.

Reviewer 2: ```suggestion
In the original speech color changer demo, we included extra lines to handle browsers that support the Web Speech API only with vendor-prefixed properties (see the [Prefixed properties](#prefixed_properties) section for more details).

In the on-device version of the demo, prefix-handling code is not needed because the implementations that support this functionality do so without prefixes.
```

> ## Speech recognition contextual biasing

Reviewer 2: or

```suggestion
## Contextual biasing in speech recognition
```

> There are times where a speech recognition service will fail to correctly recognize a specific word or phrase. This is most commonly the case for domain-specific terminology (for example medical or scientific terms), proper nouns, uncommon phrases, or words that are similar to other words and so may be misrecognized.
> 
> For example, during testing, we found that our [on-device speech color changer](https://mdn.github.io/dom-examples/web-speech-api/speech-color-changer/) had trouble recognizing the color `azure` — it kept returning results like "as you". Other colors that were frequently misrecognized included `khaki` ("car key"), `tan`, and `thistle` ("this all").
> 
> To mitigate such problems, the Web Speech API can provide hints to the recognition engine to highlight phrases that are more likely to be spoken and which the engine should therefore be biased towards. The end result is that these phrases are more likely to be recognized correctly.

Reviewer 2: ```suggestion
There will be times when a speech recognition service will fail to correctly recognize a specific word or phrase. This most often happens with domain-specific terms (such as medical or scientific vocabulary), proper nouns, uncommon phrases, or words that sound similar to other words and so may be misrecognized.

For example, during testing, we found that our [On-device speech color changer](https://mdn.github.io/dom-examples/web-speech-api/speech-color-changer/) had trouble recognizing the color `azure` — it kept returning results like "as you". Other colors that were frequently misrecognized included `khaki` ("car key"), `tan`, and `thistle` ("this all").

To mitigate such problems, the Web Speech API lets you provide hints to the recognition engine to highlight phrases that are more likely to be spoken and which the engine should be biased towards. This makes those words and phrases more likely to be recognized correctly.
```

> This is done by setting an array of {{domxref("SpeechRecognitionPhrase")}} objects as the value of the {{domxref("SpeechRecognition.phrases")}} property. Each `SpeechRecognitionPhrase` object contains:

Reviewer 2: ```suggestion
You can do this by setting an array of {{domxref("SpeechRecognitionPhrase")}} objects as the value of the {{domxref("SpeechRecognition.phrases")}} property. Each `SpeechRecognitionPhrase` object contains:
```

> - A `phrase` property, which is a string containing the word or phrase you want boost.
> - A `boost` property, which is a floating point number between `0.0` and `10.0` (inclusive) that sets the amount of boost you want to give to that word or phrase. Higher values make the word or phrase more likely to be recognized.

Reviewer 2: ```suggestion
- A `phrase` property, which is a string containing the word or phrase you want to boost.
- A `boost` property, which is a floating point number between `0.0` and `10.0` (inclusive) that sets the amount of boost you want to apply to that word or phrase. Higher values make the word or phrase more likely to be recognized.
```

> - A `phrase` property, which is a string containing a word or phrase you want boosted in terms of the engine's bias towards it.
> - A `boost` property, which is a floating pointing number between `0.0` and `10.0` (inclusive) representing the amount of boost you want to give that phrase. Higher values make the phrase more likely to be recognized.

Reviewer 2: ```suggestion
- A `phrase` property, which is a string containing the word or phrase you want boost.
- A `boost` property, which is a floating point number between `0.0` and `10.0` (inclusive) that sets the amount of boost you want to give to that word or phrase. Higher values make the word or phrase more likely to be recognized.
```

> In our `on-device speech color changer` demo, we handle this by first creating an array containing the phrases to boost and their boost values:

Reviewer 2: ```suggestion
In our "On-device speech color changer" demo, we handle this by creating an array of phrases to boost and their boost values:
```

```js
const phraseData = [
  { phrase: "azure", boost: 5.0 },
  { phrase: "khaki", boost: 3.0 },
  { phrase: "tan", boost: 2.0 },
];
```

> These need to be represented as an `ObservableArray` of `SpeechRecognitionPhrase` objects. We can handle this by mapping the original array, converting each array element into a `SpeechRecognitionPhrase` object using the {{domxref("SpeechRecognitionPhrase.SpeechRecognitionPhrase", "SpeechRecognitionPhrase()")}} constructor:

Reviewer 2: ```suggestion
These need to be represented as an `ObservableArray` of `SpeechRecognitionPhrase` objects. We handle this by mapping the original array and converting each array element into a `SpeechRecognitionPhrase` object using the {{domxref("SpeechRecognitionPhrase.SpeechRecognitionPhrase", "SpeechRecognitionPhrase()")}} constructor:
```

```js
const phraseObjects = phraseData.map(
  (p) => new SpeechRecognitionPhrase(p.phrase, p.boost),
);
```

> After we've created our `SpeechRecognition` instance, we can then plug our contextual biasing phrases into it by setting the `phraseObjects` array as the value of the `SpeechRecognition.phrases` property:

Reviewer 2: ```suggestion
After creating the `SpeechRecognition` instance, we add our contextual biasing phrases by setting the `phraseObjects` array as the value of the `SpeechRecognition.phrases` property:
```

```js
recognition.phrases = phraseObjects;
```

The phrases array can be modified just like a normal JavaScript array, for example by pushing new phrases to it dynamically:

```js
recognition.phrases.push(new SpeechRecognitionPhrase("thistle", 5.0));
```

> After adding in this code, we found that the problematic color keywords were recognized more accurately than before.

Reviewer 2: ```suggestion
With this code, we found that the problematic color keywords were recognized more accurately than before.
```

### Demo

> To show usage of Web speech synthesis, we've provided a demo called [Speak easy synthesis](https://github.com/mdn/dom-examples/tree/main/web-speech-api/speak-easy-synthesis). This includes a set of form controls for entering text to be synthesized, and setting the pitch, rate, and voice to use when the text is uttered. After you have entered your text, you can press <kbd>Enter</kbd>/<kbd>Return</kbd> to hear it spoken.

Reviewer 2: ```suggestion
To demonstrate how to use web speech synthesis, we've created a sample app called [Speech synthesiser](https://github.com/mdn/dom-examples/tree/main/web-speech-api/speak-easy-synthesis). It has an input field for entering the text to be synthesized. You can adjust the rate and pitch and also select a voice from the dropdown menu to use for the spoken text. After you've entered your text, press <kbd>Enter</kbd>/<kbd>Return</kbd> or click the **Play** button to hear the text read aloud.
```

![UI of an app called speak easy synthesis. It has an input field in which to input text to be synthesized, slider controls to change the rate and pitch of the speech, and a drop down menu to choose between different voices.](speak-easy-synthesis.png)

> To run the demo, navigate to the [live demo URL](https://mdn.github.io/dom-examples/web-speech-api/speak-easy-synthesis/) in a supporting browser.

Reviewer 2: ```suggestion
To run the demo, navigate to the [live demo URL](https://mdn.github.io/dom-examples/web-speech-api/speak-easy-synthesis/) in a [supporting browser](/en-US/docs/Web/API/SpeechSynthesis#browser_compatibility).
```

### HTML and CSS

> The HTML and CSS are again pretty trivial, containing a title, some instructions for use, and a form with some basic controls. The {{htmlelement("select")}} element is initially empty, but is populated with {{htmlelement("option")}}s via JavaScript (see later on.)

Reviewer 2: ```suggestion
The HTML and CSS for this app are pretty basic. There's a title, some instructions for use, and a form with some basic controls. The {{htmlelement("select")}} element is initially empty; it is populated with {{htmlelement("option")}}s via JavaScript (covered later).
```
